Sources/NIOHTTP1/HTTPTypes.swift

```diff
@@ -145,8 +162,33 @@ public struct HTTPResponseHead: Equatable {
     }
 }
 
-fileprivate typealias HTTPHeadersStorage = [String: [(String, String)]] // [lowerCasedName: [(originalCaseName, value)]
+/// The Index for a header name or value that points into the underlying `ByteBuffer`.
+struct HTTPHeaderIndex {
+    let start: Int
+    let length: Int
+}
+
+struct HTTPHeader {
+    let name: HTTPHeaderIndex
+    let value: HTTPHeaderIndex
+}
 
+private extension ByteBuffer {
+
+    func equalCaseInsensitiveASCII<S: Sequence>(seq: S, at: HTTPHeaderIndex) -> Bool where S.Element == UInt8 {
+        return withVeryUnsafeBytes { buffer in
+            let address = buffer.baseAddress!.assumingMemoryBound(to: UInt8.self)
+            var idx = 0
+            for byte in seq {
+                guard address.advanced(by: at.start + idx).pointee & 0xdf == byte & 0xdf else {
```

Comment:
This code mishandles memory. Specifically, it all goes to pieces if you do this:

```swift
var b = allocator.buffer(capacity: 2)
b.write(staticString: "co")
b.equalCaseInsensitiveASCII(seq: "content-length", at: HTTPHeaderIndex(start: 0, length: 14))
```

This comparison will overrun the ByteBuffer. The length comparison you do at the bottom of this function should come first, rather than last, and should also include a comparison of the actual length of the buffer (that is, `seq.length == at.length && (at.start <= self.capacity - at.length)`. This will also provide a performance improvement when searching for headers: headers that do not have the same length as the one we're searching for will now return false faster than they used to, and headers that do still needed to do that check.

Reply:
This should never happen as we have control over the `ByteBuffer`. Let me add an assert.